uefi-test-runner/src/proto/media/known_disk.rs

```diff
@@ -138,6 +143,119 @@ fn test_create_file(directory: &mut Directory) {
+        .expect("Failed to read from disk");
+
+    // Verify that the disk's MBR signature is correct
+    assert_eq!(slice[510], 0x55);
+    assert_eq!(slice[511], 0xaa);
+
+    info!("Raw disk I/O succeeded");
+    bt.free_pool(buf).unwrap();
+}
+
+/// Asynchronous disk I/O task context
+#[repr(C)]
+struct DiskIoTask {
+    /// Token for the transaction
+    token: DiskIo2Token,
+    /// Pointer to a buffer holding the read data
+    buffer: *mut u8,
+}
+
+/// Asynchronous disk I/O 2 transaction callback
+unsafe extern "efiapi" fn disk_io2_callback(event: Event, ctx: Option<NonNull<c_void>>) {
+    let task = ctx.unwrap().as_ptr() as *mut DiskIoTask;
+
+    // Verify that the disk's MBR signature is correct
+    assert_eq!((*task).token.transaction_status, uefi::Status::SUCCESS);
+    assert_eq!(*(*task).buffer.offset(510), 0x55);
+    assert_eq!(*(*task).buffer.offset(511), 0xaa);
+
+    // Close the completion event
+    let bt = system_table().as_ref().boot_services();
+    bt.close_event(event).unwrap();
+
+    // Free the disk data buffer & task context
+    bt.free_pool((*task).buffer).unwrap();
+    bt.free_pool(task as *mut u8).unwrap();
+}
+
+/// Tests raw disk I/O through the DiskIo2 protocol.
+fn test_raw_disk_io2(handle: Handle, image: Handle, bt: &BootServices) {
+    info!("Testing raw disk I/O 2");
+
+    // Open the disk I/O protocol on the input handle
+    if let Ok(disk_io2) = bt.open_protocol::<DiskIo2>(
+        OpenProtocolParams {
+            handle,
+            agent: image,
+            controller: None,
+        },
+        OpenProtocolAttributes::GetProtocol,
+    ) {
+        // Allocate the task context structure
+        let task = bt
+            .allocate_pool(MemoryType::LOADER_DATA, core::mem::size_of::<DiskIoTask>())
+            .expect("Failed to allocate task struct for disk I/O")
+            as *mut DiskIoTask;
+
+        // SAFETY: `task` refers to a valid allocation of at least `size_of::<DiskIoTask>()`
+        unsafe {
+            // Create the completion event as part of the disk I/O token
+            (*task).token.event = Some(
+                bt.create_event(
+                    EventType::NOTIFY_SIGNAL,
+                    Tpl::NOTIFY,
+                    Some(disk_io2_callback),
+                    NonNull::new(task as *mut c_void),
+                )
+                .expect("Failed to create disk I/O completion event"),
+            );
+
+            // Allocate a buffer for the transaction to read into
+            const SIZE_TO_READ: usize = 512;
+            (*task).buffer = bt
+                .allocate_pool(MemoryType::LOADER_DATA, SIZE_TO_READ)
+                .expect("Failed to allocate buffer for disk I/O");
+
+            // Initiate the asynchronous read operation
+            disk_io2
+                .read_disk_raw(0, 0, &mut (*task).token, SIZE_TO_READ, (*task).buffer)
+                .expect("Failed to initiate asynchronous disk I/O read");
+        }
+
```

Comment:
I haven't had a chance to fully review the async part yet, but one thing I think is missing is verification that the callback is actually being run -- I would expect some kind of wait or polling is needed here?

Reply:
I don't think this is possible due to how events work in UEFI. You cannot wait on `NOTIFY_SIGNAL` events and any kind of synchronous polling would block the asynchronous transaction from completing, due to a lack of yields.
I could change the test to behave fully synchronously instead by not providing a token just to verify the protocol executes as intended but that's about it as far as I can tell.

Reply:
Synchronous polling actually does end up working because UEFI uses a timer interrupt to handle event signals. To make this work with `wait_for_signal`, the trick seems to be to not set any bits in the `EventType`. I tested this locally (in combination with the `Option<Event>` -> `Event` change mentioned in another comment):

```rust
(*task).token.event = bt
    .create_event(
        EventType::empty(),
        Tpl::NOTIFY,
        Some(disk_io2_callback),
        NonNull::new(task as *mut c_void),
    )
    .expect("Failed to create disk I/O completion event");
let mut events = [(*task).token.event.unsafe_clone()];

// Allocate a buffer for the transaction to read into
const SIZE_TO_READ: usize = 512;
(*task).buffer = bt
    .allocate_pool(MemoryType::LOADER_DATA, SIZE_TO_READ)
    .expect("Failed to allocate buffer for disk I/O");

// Initiate the asynchronous read operation
disk_io2
    .read_disk_raw(
        block_io.media().media_id(),
        0,
        &mut (*task).token,
        SIZE_TO_READ,
        (*task).buffer,
    )
    .expect("Failed to initiate asynchronous disk I/O read");

bt.wait_for_event(&mut events).unwrap();
assert_eq!((*task).token.transaction_status, Status::SUCCESS);
```

Since you can then guarantee that the `test_raw_disk_io2` stack is still available, you can also avoid using `allocate_pool` here and allocate the buffer and token structures on the stack.

Reply:
Interesting, I previously tried creating a second event with type `0`, and then waiting on that second barrier event until its signalled by the disk I/O completion callback - however this approach just led to a hang.